### PR TITLE
fix(npm): reload an npm package's dependency's information when version not found

### PR DESCRIPTION
**Pull request by dsherret**
This reloads an npm package's dependency's information when a version/version req/tag is not found.

This PR applies only to dependencies of npm packages. It does NOT yet cause npm specifiers to have their dependency information cache busted. That requires a different solution, but this should help cache bust in more scenarios.

Part of #16901, but doesn't close it yet

### CODE REVIEW

File: cli/tests/integration/npm_tests.rs [hunk 1]
```diff
@@ -1546,3 +1548,124 @@ itest!(node_modules_import_check {
...
+
+  // now try running without it, it currently will error, but this should work in the future
+  // todo(https://github.com/denoland/deno/issues/16901): fix this
+  let output = test_context.new_command().args("run main.ts").run();
+  output.assert_exit_code(1);
+  output.assert_matches_text(concat!(
+    "error: Could not find npm package '@denotest/esm-import-cjs-default' matching '1.0.0'. Try retrieving the latest npm package information by running with --reload\n",
+    "    at file:///[WILDCARD]/main.ts:1:8\n",
+  ));
+}
```
Comment (bartlomieju) [inline cli/tests/integration/npm_tests.rs:1671]: Great test, I really like the reviewing them with the new test infrastructure. So much easier to understand without the boilerplate.

File: cli/npm/resolution.rs [hunk 2]
```diff
@@ -188,7 +191,7 @@ impl NpmResolution {
   pub fn resolve_pkg_id_from_deno_module(
     &self,
     id: &NpmPackageNv,
-  ) -> Result<NpmPackageId, AnyError> {
+  ) -> Result<NpmPackageId, PackageNvNotFoundError> {
```
Comment (bartlomieju) [inline cli/npm/resolution.rs:194]: Slowly grinding to strongly typed errors 💪

**Review by bartlomieju — APPROVED** [review]
LGTM, this change is so welcome

File: cli/args/lockfile.rs [hunk 3]
```diff
@@ -117,10 +117,11 @@ pub async fn snapshot_from_lockfile(
...
         |nv| async move {
-          match api.package_version_info(&nv).await? {
-            Some(version_info) => Ok(version_info),
-            None => {
-              bail!("could not find '{}' specified in the lockfile. Maybe try again with --reload", nv);
+          let package_info = api.package_info(&nv.name).await?;
+          match package_info.version_info(&nv) {
+            Ok(version_info) => Ok(version_info),
+            Err(err) => {
+              bail!("Could not find '{}' specified in the lockfile. Maybe try again with --reload", err.0);
```
Comment (dsherret) [inline cli/args/lockfile.rs:124]: Oh, we should update this to also not require `--reload`. I will do it in a future PR.

File: cli/tests/integration/npm_tests.rs (hunk 1 above)
Comment (dsherret) [reply cli/tests/integration/npm_tests.rs:1671]: It's less arduous to write too!

File: cli/npm/resolution.rs [hunk 4]
```diff
@@ -257,9 +264,23 @@ async fn add_package_reqs_to_snapshot(
...
+    Err(NpmResolutionError::Resolution(err)) if api.mark_force_reload() => {
+      log::debug!("{err:#}");
+      log::debug!("npm resolution failed. Trying again...");
+
+      // try again
+      let snapshot = get_new_snapshot();
+      let result = snapshot.resolve_pending(package_reqs, api).await;
+      api.clear_memory_cache();
+      // now surface the result after clearing the cache
+      result?
```
Comment (dsherret) [inline cli/npm/resolution.rs:280]: One thing I should have done is just surface the original error here. I didn't think about that until now. I will do that in a future PR.

Comment (dsherret) [reply cli/npm/resolution.rs:280]: Actually, this might be better.

**Pull request merged**